jax/_src/lax/lax.py

```diff
@@ -1330,11 +1338,11 @@ def full_like(x: Array, fill_value: ArrayLike, dtype: Optional[DTypeLike] = None
   # probably in the form of a primitive like `val = match_sharding_p.bind(x, val)`
   # (so it works in staged-out code as well as 'eager' code). Related to
   # equi-sharding.
-  if (config.jax_array and hasattr(x, 'sharding') and
-      not dispatch.is_single_device_sharding(x.sharding) and
-      not isinstance(x.sharding, sharding.PmapSharding)):
-    return array.make_array_from_callback(
-        fill_shape, x.sharding, lambda idx: val[idx])  # type: ignore[arg-type]
+  if config.jax_array and hasattr(x, 'sharding'):
+    sharding = x.sharding  # type: ignore[union-attr]
```

Comment:
Why this change? Do I have to use this pattern everywhere?

Reply:
Because `hasattr(x, 'sharding')` is not enough for mypy to understand that `x` is now of a type that has an attribute `sharding`, so `x.sharding` results in a type check error. We could avoid this if every Array-like object declared a `sharding` attribute (e.g. `Tracer` would have to define sharding), or by using an `isinstance` check in place of the `hasattr` check (i.e. avoiding duck-typing) but I did it this way in order to not change the semantics of the function, which I admit I don't totally understand.

Reply:
Hmm.. okay. We will eventually end up defining a sharding attributed on Tracer too I think but let's deal with that later.